075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+In terms of scaling, the reconcile loop has the scale down triggered before the scaling up but it could be skipped because the brokers to be removed are still hosting topics' partitions. In such case the scaling up runs before.
+In terms of rebalancing, after the scaling up happens, the rebalancing for brokers to be removed takes the precedence on the rebalancing for newly added brokers. They are executed sequentially and not in parallel.
+
+### Auto-rebalancing Finite State Machine (FSM)
+
+The auto-rebalancing mechanism runs through a Finite State Machine (FSM) made by the following states:
+
+* **Idle**: Initial state with a new auto-rebalancing going to run because scaling down/up operations were requested. This is also the ending state after an auto-rebalancing completed successfully or failed.
+* **RebalanceOnScaleDown**: a rebalancing related to a scale down operation is running.
+* **RebalanceOnScaleUp**: a rebalancing related to a scale up operation is running.
+
+The FSM states' transitions are the following:
+
+* from **Idle** to:
+  * **RebalanceOnScaleDown**: if a scale down operation was requested. This transition happens even if a scale up was requested at the same time but the rebalancing on scaling down has the precedence. The rebalancing on scale up is queued. They will run sequentially one after the other.
+  * **RebalanceOnScaleUp**: if only a scale up operation was requested. There was no scale down operation requested.
+* from **RebalanceOnScaleDown** to:
+  * **RebalanceOnScaleDown**: if a rebalancing on scale down is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleUp**: if a scale down operation was requested together with a scale up and, because they run sequentially, the rebalance on scale down had the precedence, was executed first and completed successfully. We can now move on with rebalancing for the scale up.
+  * **Idle**: if only a scale down operation was requested, it was executed and completed successfully or failed.
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended. 
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
+  * **Idle**: if a scale up operation was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the logic will be like this:
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance and it includes:
+
+   * checking the scale down (if brokers cannot be removed because they host topics' partitions) and reverting replicas count if it fails.
+   * collecting removed (scale down) and added nodes (scale up). Let's call these lists as `removedNodes` and `addedNodes`.
```

Comment:
I assume you understand it well, but I think the way it is written is not completely clear -> you really case only about the reverted scale-down nodes. Not about the node passed for scaledown.

```suggestion
   * collecting the nodes reverted from scale-down (scale down) and the nodes added by scale up. Let's call these lists as `toBeRemovedNodes` and `addedNodes`.
```

Reply:
Yeah good suggestion, thanks.